geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/test/scala/org/locationtech/geomesa/fs/storage/common/s3/S3VisibilityObserverTest.scala

```diff
@@ -0,0 +1,148 @@
+    override def mkdirs(path: Path, fsp: FsPermission): Boolean = ???
+    override def getFileStatus(path: Path): FileStatus = ???
+  }
+
+  "S3VisibilityObserver" should {
+
+    "initialize factory correctly" >> {
+      // mimic construction through reflection
+      WithClose(classOf[S3VisibilityObserverFactory].newInstance()) { factory =>
+        val root = mock[Path]
+        root.getFileSystem(ArgumentMatchers.any()) returns new MockFileSystem()
+        factory.init(new Configuration(), root, sft) must not(throwAn[Exception])
+      }
+    }
+
+    "tag a single visibility label" >> {
+      WithClose(new S3VisibilityObserverFactory) { factory =>
+        val fs = new MockFileSystem()
+        val root = mock[Path]
+        root.getFileSystem(ArgumentMatchers.any()) returns fs
+        factory.init(new Configuration(), root, sft)
+        val observer = factory.apply(new Path("s3a://foo/bar/baz.json"))
+        observer.write(feature(0, "user"))
+        observer.close()
+
+        val captor: ArgumentCaptor[SetObjectTaggingRequest] = ArgumentCaptor.forClass(classOf[SetObjectTaggingRequest])
+        there was one(fs.s3).setObjectTagging(captor.capture())
+        val request = captor.getValue
+        request.getBucketName mustEqual "foo"
+        request.getKey mustEqual "bar/baz.json"
+        val encoded = Base64.getEncoder.encodeToString("user".getBytes(StandardCharsets.UTF_8))
+        request.getTagging.getTagSet.asScala mustEqual Seq(new Tag(S3VisibilityObserverFactory.DefaultTag, encoded))
+      }
+    }
+
+    "tag multiple visibility labels" >> {
+      WithClose(new S3VisibilityObserverFactory) { factory =>
+        val fs = new MockFileSystem()
+        val root = mock[Path]
+        root.getFileSystem(ArgumentMatchers.any()) returns fs
+        factory.init(new Configuration(), root, sft)
+        val observer = factory.apply(new Path("s3a://foo/bar/baz.json"))
+        observer.write(feature(0, "user"))
+        observer.write(feature(1, "admin"))
+        observer.write(feature(2, "user"))
+        observer.close()
+
+        val captor: ArgumentCaptor[SetObjectTaggingRequest] = ArgumentCaptor.forClass(classOf[SetObjectTaggingRequest])
+        there was one(fs.s3).setObjectTagging(captor.capture())
+        val request = captor.getValue
+        request.getBucketName mustEqual "foo"
+        request.getKey mustEqual "bar/baz.json"
+        // since the vis are kept in a set, the order is not defined
+        val encodedFront = Base64.getEncoder.encodeToString("user&admin".getBytes(StandardCharsets.UTF_8))
+        val encodedBack = Base64.getEncoder.encodeToString("admin&user".getBytes(StandardCharsets.UTF_8))
+        request.getTagging.getTagSet.asScala must beOneOf(
+          Seq(new Tag(S3VisibilityObserverFactory.DefaultTag, encodedFront)),
+          Seq(new Tag(S3VisibilityObserverFactory.DefaultTag, encodedBack))
+        )
+      }
+    }
+
+    "simplify tag expressions" >> {
```

Comment:
What's does the simplification?  

Does it handle complicated conjunctions and disjunctions?

Reply:
the simplification is here: https://github.com/locationtech/geomesa/pull/2453/files/9b4cf3563a078f7aeb687586b1d9d92a2a52ca0b#diff-25df5de0778453f8b01569cccd7efd95R40-R46
It's not that complicated, it just flattens AND clauses and deduplicates. so e.g. `foo&bar` plus `foo` ends up as `foo&bar` and not as `(foo&bar)&foo`